### DISCUSSION TITLE
Creating SWR using futureprovider

### DISCUSSION THREAD

**Discussion by kyeshmz** [rrousselGit/riverpod#1563] [Q&A]
Hi!

I have a usecase where I need to update a screen as frequently as possible for database changes and synchronization.
I was wondering if some sort of polling could be done, while also utilizing some of the caching that future provider has already.
In a nutshell, I was wondering if someone has already done something like [Vercel's SWR](https://swr.vercel.app/), but for flutter and with riverpod?

I saw some  other people in other state management libraries try to figure this out 
https://github.com/felangel/bloc/issues/477
```


Stream<AppConnectionState> _startConnectionPolling() async* {
    timer = Timer.periodic(Duration(seconds: 10), (timer) async {
      var apiRes = await apiProvider.retrieveDetails();
      if (apiRes.connected)
        dispatch(Connected());
      else
        dispatch(Disconnected());
    });
  }

  @override
  void dispose() {
    timer?.cancel();
    super.dispose();
  }
```

**Comment by rrousselGit** [comment]
Polling is simple to do in Riverpod, by making a StreamProvider thar uses Stream.periodic

Then you can what that provider I side other providers.
This will cause them to recomputed whenever the period ticks

**Comment by kyeshmz** [comment]
Thank you for the tip!
I was able to do some kind of polling utilizing future provider, I was wondering if 
1. Should you be caching these kind of calls into a state notifier?
2. Is there a way to manually call an update to a stream ?
ie. make a stream "get" some kind of api call before the time period?

```

  Future<List<LogDB>> getItemLogs(String id) async {
    final url =
        Uri.parse('${dotenv.env['API_URL']}/api/v1/item/log?itemId=$id');

    final res = await http.get(url);

    final data = jsonDecode(res.body);
    List<LogDB> returnItems = [];
    for (var element in (data['message'] as List<dynamic>)) {
      final tmp = LogDB.fromJson(element);
      returnItems.add(tmp);
    }
    return returnItems;
  }

  Stream<List<LogDB>> getItemLogStream(String id) =>
      Stream.periodic(const Duration(seconds: 3))
          .asyncMap((_) => getItemLogs(id));
```

**Reply by rrousselGit** [reply]
You can use ref.invslidate on your provider that does the request to trigger a new one.

**Reply by kyeshmz** [reply]
Ah, thank you so much! Will think about caching next

**Reply by rrousselGit** [reply]
Caching is already taken care of my providers.

There are some features that I'd like to add, but overall Riverpod does quite a lot for swr already

**Reply by rrousselGit** [reply]
Caching is already taken care of my providers.

There are some features that I'd like to add, but overall Riverpod does quite a lot for swr already
